src/browser/cached_resource_fetcher.ts

```
@@ -13,85 +13,213 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-                            });
-                        }
+                        reject(err);
                     } else {
-                        download(fileUrl, filePath, callback);
+                        resolve(appCacheDir);
                     }
                 });
+            } else {
+                resolve(appCacheDir);
             }
-        } else if (remoteFileIsYoungerThanCachedFile(fileUrl, filePath)) {
-            download(fileUrl, filePath, callback);
-        } else {
-            callback(null, filePath);
-        }
+        });
     });
 }
 
-function remoteFileIsYoungerThanCachedFile(remoteUrl: string, cachedFilePath: string) {
-    //todo: make a RESTful HEAD request and if file at remoteUrl is missing, return false;
-    //todo: else if file at remoteUrl is younger than file at cachedFilePath, return true;
-    //todo: else return false
-    return true; //for now we are always fetching
+const cachedFiles: { [key: string]: Promise<string> } = {};
+
+function getFile(dirPath: string, url: string, data: string | Buffer, encoding: string): Promise<string> {
+    const filePath: string = getFilePath(dirPath, url);
+    return cachedFiles[filePath] ?
+        Promise.resolve(cachedFiles[filePath]) :
+        (cachedFiles[filePath] = write(filePath, data, encoding));
+}
+
+function write(filePath: string, data: string | Buffer, encoding: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+        stat(filePath, (err: Error) => {
+            if (err) {
+                writeFile(filePath, data, {encoding}, (err: Error) => {
+                    if (err) {
+                        reject(err);
+                    } else {
+                        resolve(filePath);
+                    }
+                });
+            } else {
+                resolve(filePath);
+            }
+        });
+    });
+}
+
+function createFromPath(filePath?: string, expires: number = -1): Promise<NativeImage> {
```

This function does a lot more than `createFromPath`.... also is a silent fail what we really want?

---

the file delete failure unexpected and is not something we can do anything about. is it worth logging?